snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py

```diff
@@ -0,0 +1,131 @@
+from snuba.processor import MAX_UINT32, NIL_UUID
+from .matview import create_matview_v1
+
+
+aggregate_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("started", DateTime()),
+    Column("release", String(Modifiers(low_cardinality=True))),
+    Column("environment", String(Modifiers(low_cardinality=True))),
+    Column("user_agent", String(Modifiers(low_cardinality=True))),
+    Column("os", String(Modifiers(low_cardinality=True))),
+    # durations
+    Column(
+        "duration_quantiles",
+        AggregateFunction("quantilesIf(0.5, 0.9)", [UInt(32), UInt(8)]),
+    ),
+    Column("duration_avg", AggregateFunction("avgIf", [UInt(32), UInt(8)])),
+    # sessions:
+    Column("sessions", AggregateFunction("countIf", [UUID(), UInt(8)])),
+    Column("sessions_preaggr", AggregateFunction("sumIf", [UInt(32), UInt(8)])),
+    Column("sessions_crashed", AggregateFunction("countIf", [UUID(), UInt(8)])),
+    Column("sessions_crashed_preaggr", AggregateFunction("sumIf", [UInt(32), UInt(8)])),
+    Column("sessions_abnormal", AggregateFunction("countIf", [UUID(), UInt(8)])),
+    Column(
+        "sessions_abnormal_preaggr", AggregateFunction("sumIf", [UInt(32), UInt(8)])
+    ),
+    Column("sessions_errored", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+    Column("sessions_errored_preaggr", AggregateFunction("sumIf", [UInt(32), UInt(8)])),
+    # users:
+    Column("users", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+    Column("users_crashed", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+    Column("users_abnormal", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+    Column("users_errored", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+]
+
+
+query = f"""
+SELECT
+    org_id,
+    project_id,
+    toStartOfHour(started) as started,
+    release,
+    environment,
+    user_agent,
+    os,
+
+    -- pre-aggregated sessions dont have a duration, so no more filtering is needed.
+    -- we would have liked to changed the quantiles here to `0.5, 0.75, 0.95, 0.99`:
+    quantilesIfState(0.5, 0.9)(
+        duration,
+        duration <> {MAX_UINT32} AND status == 1
+    ) as duration_quantiles,
+
+    -- this is new:
+    avgIfState(duration, duration <> {MAX_UINT32} AND status == 1) as duration_avg,
+
+    -- `sum` the session counts based on the new `quantity`:
+    sumIfState(quantity, seq == 0) as sessions_preaggr,
+    sumIfState(quantity, status == 2) as sessions_crashed_preaggr,
+    sumIfState(quantity, status == 3) as sessions_abnormal_preaggr,
+
+    -- individual session updates keep using the uniq session_id as before:
+    uniqIfState(session_id, errors > 0 AND session_id != '{NIL_UUID}') as sessions_errored,
+
+    -- pre-aggregated counts use sum. by definition, `crashed` and `abnormal` are errored:
+    sumIfState(quantity, status IN (2, 3, 4) AND session_id == '{NIL_UUID}') as sessions_errored_preaggr,
```

Comment:
Is there any session coming in with session_id == NIL_UUID as of now that would be counted in this counter ?

Reply:
Relay will enforce a uuid for individual session updates.